tensorpipe/common/cuda_host_allocator.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/common/cuda_host_allocator.h>
+
+#include <cuda_runtime.h>
+
+#include <tensorpipe/common/cuda.h>
+
+namespace tensorpipe {
+
+CudaHostAllocator::CudaHostAllocator(size_t numChunks, size_t chunkSize)
+    : numChunks_(numChunks),
+      chunkSize_(chunkSize),
+      data_(allocPinnedBuffer(numChunks * chunkSize)),
+      chunkAvailable_(numChunks, true) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  thread_ = std::thread([this, lock{std::move(lock)}]() mutable {
+    setThreadName("TP_CUDA_host_allocator_loop");
+    allocLoop(std::move(lock));
```

Comment:
You had used this pattern before (acquire lock in constructor, move into main loop thread), and I think we concluded that it wasn't necessary. I've got the same feeling now, as you sure this is needed?

The reason for my feeling is that the main thread will inevitably end up releasing and then re-acquiring the lock at some point, typically during wait, and thus needs to be able to determine on its own whether it is able to perform some work or not (in particular, it needs to do so before deciding to go to sleep). Hence that thread should be able to cope with the fact of starting up without already holding the lock.